File: riveredge-backend/src/tree_root/schemas/code_rule.py
"""
编码规则 Schema 模块

定义编码规则相关的 Pydantic Schema，用于 API 请求和响应验证。
"""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, ConfigDict, field_validator


class CodeRuleBase(BaseModel):
    """
    编码规则基础 Schema
    
    包含编码规则的基本字段，用于创建和更新操作。
    """
    name: str = Field(..., min_length=1, max_length=100, description="规则名称")
    code: str = Field(..., min_length=1, max_length=50, description="规则代码（唯一，用于程序识别）")
    expression: str = Field(..., min_length=1, max_length=200, description="规则表达式")
    description: Optional[str] = Field(None, description="规则描述")
    seq_start: int = Field(default=1, ge=0, description="序号起始值")
    seq_step: int = Field(default=1, ge=1, description="序号步长")
    seq_reset_rule: Optional[str] = Field(None, description="序号重置规则：never、daily、monthly、yearly")
    is_system: bool = Field(default=False, description="是否系统规则")
    is_active: bool = Field(default=True, description="是否启用")
    
    @field_validator("seq_reset_rule")
    @classmethod
    def validate_reset_rule(cls, v: Optional[str]) -> Optional[str]:
        """
        验证序号重置规则
        
        Args:
            v: 重置规则值
            
        Returns:
            验证后的重置规则值
            
        Raises:
            ValueError: 如果重置规则不合法
        """
        if v is not None and v not in ("never", "daily", "monthly", "yearly"):
            raise ValueError("序号重置规则必须是 never、daily、monthly 或 yearly")
        return v


class CodeRuleCreate(CodeRuleBase):
    """
    编码规则创建 Schema
    
    用于创建新编码规则的请求数据。
    """
    pass


class CodeRuleUpdate(BaseModel):
    """
    编码规则更新 Schema
    
    用于更新编码规则的请求数据，所有字段可选。
    """
    name: Optional[str] = Field(None, min_length=1, max_length=100, description="规则名称")
    expression: Optional[str] = Field(None, min_length=1, max_length=200, description="规则表达式")
    description: Optional[str] = Field(None, description="规则描述")
    seq_start: Optional[int] = Field(None, ge=0, description="序号起始值")
    seq_step: Optional[int] = Field(None, ge=1, description="序号步长")
    seq_reset_rule: Optional[str] = Field(None, description="序号重置规则：never、daily、monthly、yearly")
    is_active: Optional[bool] = Field(None, description="是否启用")
    
    @field_validator("seq_reset_rule")
    @classmethod
    def validate_reset_rule(cls, v: Optional[str]) -> Optional[str]:
        """
        验证序号重置规则
        
        Args:
            v: 重置规则值
            
        Returns:
            验证后的重置规则值
            
        Raises:
            ValueError: 如果重置规则不合法
        """
        if v is not None and v not in ("never", "daily", "monthly", "yearly"):
            raise ValueError("序号重置规则必须是 never、daily、monthly 或 yearly")
        return v


class CodeRuleResponse(CodeRuleBase):
    """
    编码规则响应 Schema
    
    用于返回编码规则信息。
    """
    uuid: str = Field(..., description="规则UUID（对外暴露，业务标识）")
    tenant_id: int = Field(..., description="组织ID")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    
    model_config = ConfigDict(from_attributes=True)


class CodeGenerationRequest(BaseModel):
    """
    编码生成请求 Schema
    
    用于测试编码生成功能。
    """
    rule_code: str = Field(..., description="规则代码")
    context: Optional[dict] = Field(None, description="上下文变量（可选，用于自定义变量）")


class CodeGenerationResponse(BaseModel):
    """
    编码生成响应 Schema
    
    用于返回生成的编码。
    """
    code: str = Field(..., description="生成的编码")
    rule_name: str = Field(..., description="规则名称")

File: riveredge-frontend/src/services/codeRule.ts
/**
 * 编码规则 API 服务
 * 
 * 提供编码规则管理相关的 API 接口
 * 注意：所有 API 自动过滤当前组织的编码规则
 */

import { apiRequest } from './api';

/**
 * 编码规则信息接口
 */
export interface CodeRule {
  uuid: string;
  name: string;
  code: string;
  expression: string;
  description?: string;
  seq_start: number;
  seq_step: number;
  seq_reset_rule?: 'never' | 'daily' | 'monthly' | 'yearly';
  is_system: boolean;
  is_active: boolean;
  allow_manual_edit?: boolean;
  tenant_id: number;
  created_at: string;
  updated_at: string;
}

/**
 * 编码规则列表查询参数
 */
export interface CodeRuleListParams {
  page?: number;
  page_size?: number;
  is_active?: boolean;
}

/**
 * 编码规则列表响应数据
 */
export interface CodeRuleListResponse {
  items: CodeRule[];
  total: number;
}

/**
 * 创建编码规则数据
 */
export interface CreateCodeRuleData {
  name: string;
  code: string;
  expression: string;
  description?: string;
  seq_start?: number;
  seq_step?: number;
  seq_reset_rule?: 'never' | 'daily' | 'monthly' | 'yearly';
  is_system?: boolean;
  is_active?: boolean;
}

/**
 * 更新编码规则数据
 */
export interface UpdateCodeRuleData {
  name?: string;
  expression?: string;
  description?: string;
  seq_start?: number;
  seq_step?: number;
  seq_reset_rule?: 'never' | 'daily' | 'monthly' | 'yearly';
  is_active?: boolean;
}

/**
 * 编码生成请求数据
 */
export interface CodeGenerationRequest {
  rule_code: string;
  context?: Record<string, any>;
  check_duplicate?: boolean; // 是否检查重复（如果为True，会自动递增直到找到不重复的编码）
  entity_type?: string; // 实体类型（如：'material'，用于检查重复）
}

/**
 * 编码生成响应数据
 */
export interface CodeGenerationResponse {
  code: string;
  rule_name: string;
}

/**
 * 获取编码规则列表
 * 
 * 自动过滤当前组织的编码规则。
 * 注意：后端返回的是 List[CodeRuleResponse]（数组），需要转换为分页格式。
 * 
 * @param params - 查询参数
 * @returns 编码规则列表响应数据
 */
export async function getCodeRuleList(params?: CodeRuleListParams): Promise<CodeRuleListResponse> {
  // 后端 API 使用 skip 和 limit 参数，而不是 page 和 page_size
  const queryParams: any = {};
  if (params?.page && params?.page_size) {
    queryParams.skip = (params.page - 1) * params.page_size;
    queryParams.limit = params.page_size;
  }
  if (params?.is_active !== undefined) {
    queryParams.is_active = params.is_active;
  }
  
  // 后端返回的是数组，需要转换为分页格式
  const response = await apiRequest<CodeRule[]>('/core/code-rules', {
    params: queryParams,
  });
  
  // 如果是数组，转换为分页格式
  if (Array.isArray(response)) {
    return {
      items: response,
      total: response.length,
    };
  }
  
  // 如果已经是分页格式，直接返回
  if (response && typeof response === 'object' && 'items' in response) {
    return response as CodeRuleListResponse;
  }
  
  // 默认返回空列表
  return {
    items: [],
    total: 0,
  };
}

/**
 * 获取编码规则详情
 * 
 * 自动验证组织权限：只能获取当前组织的编码规则。
 * 
 * @param ruleUuid - 编码规则 UUID
 * @returns 编码规则信息
 */
export async function getCodeRuleByUuid(ruleUuid: string): Promise<CodeRule> {
  return apiRequest<CodeRule>(`/core/code-rules/${ruleUuid}`);
}

/**
 * 创建编码规则
 * 
 * 自动设置当前组织的 tenant_id。
 * 
 * @param data - 编码规则创建数据
 * @returns 创建的编码规则信息
 */
export async function createCodeRule(data: CreateCodeRuleData): Promise<CodeRule> {
  return apiRequest<CodeRule>('/core/code-rules', {
    method: 'POST',
    data,
  });
}

/**
 * 更新编码规则
 * 
 * 自动验证组织权限：只能更新当前组织的编码规则。
 * 
 * @param ruleUuid - 编码规则 UUID
 * @param data - 编码规则更新数据
 * @returns 更新后的编码规则信息
 */
export async function updateCodeRule(ruleUuid: string, data: UpdateCodeRuleData): Promise<CodeRule> {
  return apiRequest<CodeRule>(`/core/code-rules/${ruleUuid}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除编码规则
 * 
 * 自动验证组织权限：只能删除当前组织的编码规则。
 * 系统规则不可删除。
 * 
 * @param ruleUuid - 编码规则 UUID
 */
export async function deleteCodeRule(ruleUuid: string): Promise<void> {
  return apiRequest<void>(`/core/code-rules/${ruleUuid}`, {
    method: 'DELETE',
  });
}

/**
 * 生成编码（会更新序号）
 * 
 * @param request - 编码生成请求数据
 * @returns 生成的编码
 */
export async function generateCode(request: CodeGenerationRequest): Promise<CodeGenerationResponse> {
  return apiRequest<CodeGenerationResponse>('/core/code-rules/generate', {
    method: 'POST',
    data: request,
  });
}

/**
 * 测试生成编码（不更新序号）
 * 
 * @param request - 编码生成请求数据
 * @returns 生成的编码（测试用）
 */
export async function testGenerateCode(request: CodeGenerationRequest): Promise<CodeGenerationResponse> {
  return apiRequest<CodeGenerationResponse>('/core/code-rules/test-generate', {
    method: 'POST',
    data: request,
  });
}

/**
 * 编码规则页面配置接口（后端返回格式，snake_case）
 */
export interface CodeRulePageConfigResponse {
  page_code: string;
  page_name: string;
  page_path: string;
  code_field: string;
  code_field_label: string;
  module: string;
  module_icon?: string;
  auto_generate?: boolean;
  rule_code?: string;
  allow_manual_edit?: boolean;
  available_fields?: Array<{
    field_name: string;
    field_label: string;
    field_type: string;
    description?: string;
  }>;
}

/**
 * 编码规则页面配置接口（前端使用格式，camelCase）
 */
export interface CodeRulePageConfig {
  pageCode: string;
  pageName: string;
  pagePath: string;
  codeField: string;
  codeFieldLabel: string;
  module: string;
  moduleIcon?: string;
  autoGenerate?: boolean;
  ruleCode?: string;
  allowManualEdit?: boolean;
  availableFields?: Array<{
    fieldName: string;
    fieldLabel: string;
    fieldType: string;
    description?: string;
  }>;
}

/**
 * 获取编码规则功能页面配置列表
 * 
 * 返回系统中所有有编码字段的功能页面配置。
 * 
 * @returns 功能页面配置列表（已转换为camelCase格式）
 */
export async function getCodeRulePages(): Promise<CodeRulePageConfig[]> {
  const pages = await apiRequest<CodeRulePageConfigResponse[]>('/core/code-rules/pages');
  // 转换字段名从 snake_case 到 camelCase
  return pages.map(page => ({
    pageCode: page.page_code,
    pageName: page.page_name,
    pagePath: page.page_path,
    codeField: page.code_field,
    codeFieldLabel: page.code_field_label,
    module: page.module,
    moduleIcon: page.module_icon,
    autoGenerate: page.auto_generate ?? false,
    ruleCode: page.rule_code,
    allowManualEdit: page.allow_manual_edit ?? true,
    availableFields: page.available_fields?.map(field => ({
      fieldName: field.field_name,
      fieldLabel: field.field_label,
      fieldType: field.field_type,
      description: field.description,
    })),
  }));
}

/**
 * 获取指定页面的编码规则配置
 * 
 * 根据页面代码获取编码规则配置，包括是否自动生成、是否允许手动填写等。
 * 
 * @param pageCode - 页面代码（如：kuaizhizao-sales-order）
 * @returns 页面编码规则配置（已转换为camelCase格式）
 */
export async function getCodeRulePageConfig(pageCode: string): Promise<CodeRulePageConfig | null> {
  try {
    const page = await apiRequest<CodeRulePageConfigResponse>(`/core/code-rules/pages/${pageCode}`);
    // 转换字段名从 snake_case 到 camelCase
    return {
      pageCode: page.page_code,
      pageName: page.page_name,
      pagePath: page.page_path,
      codeField: page.code_field,
      codeFieldLabel: page.code_field_label,
      module: page.module,
      moduleIcon: page.module_icon,
      autoGenerate: page.auto_generate ?? false,
      ruleCode: page.rule_code,
      allowManualEdit: page.allow_manual_edit ?? true,
      availableFields: page.available_fields?.map(field => ({
        fieldName: field.field_name,
        fieldLabel: field.field_label,
        fieldType: field.field_type,
        description: field.description,
      })),
    };
  } catch (error) {
    console.error('获取页面编码规则配置失败:', error);
    return null;
  }
}
